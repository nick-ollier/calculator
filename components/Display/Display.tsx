import classnames from 'classnames';

export interface DisplayProps {
    locale?: string;
    displayValue?: string;
    portrait: boolean;
}

const Display: React.FC<DisplayProps> = ({
    locale = 'en-AU',
    displayValue = '0',
    portrait
}) => {
    let value = Intl.NumberFormat(locale, {
        maximumFractionDigits: 10
    }).format(Number(displayValue));

    if (portrait && displayValue.length >= 10) {
        value = parseFloat(value).toExponential(1);
    }

    return (
        <div
            className={classnames('w-full text-right text-4xl px-5', {
                'text-6xl': value.length <= 7 && portrait,
                'text-5xl': value.length > 7 && portrait,
                'px-8': !portrait
            })}
            data-testid='display'
        >
            {value}
        </div>
    );
};

export default Display;
