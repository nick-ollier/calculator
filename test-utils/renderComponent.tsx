import CalculatorProvider from '@context/CalculatorContext';
import { render } from '@testing-library/react';

const renderWrapper = (ui: React.ReactElement) => {
    const WithProviders = ({ children }) => {
        return <CalculatorProvider>{children}</CalculatorProvider>;
    };

    return render(ui, { wrapper: WithProviders });
};

export * from '@testing-library/react';
export { renderWrapper as render };
