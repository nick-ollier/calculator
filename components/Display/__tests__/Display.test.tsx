import { render, screen } from '@testUtils/renderComponent';
import Display, { DisplayProps } from '@components/Display/Display';

const renderComponent = (props?: Partial<DisplayProps>) => {
    const utils = render(<Display portrait {...props} />);

    return { ...utils };
};

describe('component: display', () => {
    describe('portrait', () => {
        describe('default', () => {
            it('renders', () => {
                const { container } = renderComponent();
                expect(container);
            });

            it('matches snapshot', () => {
                const { asFragment } = renderComponent();
                expect(asFragment()).toMatchSnapshot();
            });

            it('renders with default value: 0', () => {
                renderComponent();
                expect(screen.getByText('0')).toBeInTheDocument();
            });

            it('renders with class: text-6xl', () => {
                renderComponent();
                expect(screen.getByText('0')).toHaveClass('text-6xl');
            });
        });

        describe('with props', () => {
            it('renders', () => {
                const { container } = renderComponent({
                    displayValue: 12345678
                });
                expect(container);
            });

            it('matches snapshot', () => {
                const { asFragment } = renderComponent({
                    displayValue: 12345678
                });
                expect(asFragment()).toMatchSnapshot();
            });

            it('displays exponential number when displayValue length is greater or equal to 10', () => {
                renderComponent({ displayValue: 1234567890 });
                expect(screen.getByText('1.0e+0')).toBeInTheDocument();
            });

            it("renders with class: text-5xl when displayValue's length is > 7", () => {
                renderComponent({ displayValue: 12345678 });
                expect(screen.getByText('12,345,678')).toHaveClass('text-5xl');
                expect(screen.getByText('12,345,678')).not.toHaveClass(
                    'text-6xl'
                );
            });
        });
    });

    describe('landscape', () => {
        describe('default', () => {
            it('renders', () => {
                const { container } = renderComponent({ portrait: false });
                expect(container);
            });

            it('matches snapshot', () => {
                const { asFragment } = renderComponent({ portrait: false });
                expect(asFragment()).toMatchSnapshot();
            });

            it('renders with default value: 0', () => {
                renderComponent({ portrait: false });
                expect(screen.getByText('0')).toBeInTheDocument();
            });

            it('renders with class: text-4xl', () => {
                renderComponent({ portrait: false });
                expect(screen.getByText('0')).toHaveClass('text-4xl');
                expect(screen.getByText('0')).not.toHaveClass('text-5xl');
            });
        });

        describe('with props', () => {
            it('renders', () => {
                const { container } = renderComponent({
                    portrait: false,
                    displayValue: 12345678
                });
                expect(container);
            });

            it('matches snapshot', () => {
                const { asFragment } = renderComponent({
                    portrait: false,
                    displayValue: 12345678
                });
                expect(asFragment()).toMatchSnapshot();
            });

            it('displays correctly formatted number when displayValue length is greater or equal to 10', () => {
                renderComponent({
                    portrait: false,
                    displayValue: 1234567890
                });
                expect(screen.getByText('1,234,567,890')).toBeInTheDocument();
            });

            it('renders with class: text-4xl regardless of displayValue length', () => {
                renderComponent({ portrait: false, displayValue: 12345678 });
                expect(screen.getByText('12,345,678')).toHaveClass('text-4xl');
                expect(screen.getByText('12,345,678')).not.toHaveClass(
                    'text-6xl'
                );
            });
        });
    });
});
