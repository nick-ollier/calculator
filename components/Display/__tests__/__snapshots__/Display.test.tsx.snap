// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: display landscape default matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-full text-right text-4xl px-5 px-8"
    data-testid="display"
  >
    0
  </div>
</DocumentFragment>
`;

exports[`component: display landscape with props matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-full text-right text-4xl px-5 px-8"
    data-testid="display"
  >
    12,345,678
  </div>
</DocumentFragment>
`;

exports[`component: display portrait default matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-full text-right text-4xl px-5 text-6xl"
    data-testid="display"
  >
    0
  </div>
</DocumentFragment>
`;

exports[`component: display portrait with props matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-full text-right text-4xl px-5 text-5xl"
    data-testid="display"
  >
    12,345,678
  </div>
</DocumentFragment>
`;
